### ISSUE TITLE
Only authorized users should be able to edit embargoes

### ISSUE THREAD

**Issue by maxkadel** [samvera/hyrax#4782]
### Descriptive summary
Hyrax v2.9.3

Currently the embargoes_controller_behavior does not authorize the user on the edit method.

### Rationale

If a user does not have permissions (via CanCanCan) to edit an embargo, they can still go directly to the edit route and edit the embargo. 

### Expected behavior
* A user who is not authorized to edit `Hydra::AccessControls::Embargo` 
  * cannot access the `/embargoes/#{work.id}/edit` path
  * cannot update a work's embargo

### Actual behavior
* A user who is not authorized to edit `Hydra::AccessControls::Embargo` 
  * can access the `/embargoes/#{work.id}/edit` path
  * can update a work's embargo

### Steps to reproduce the behavior

1. WIP
2. WIP

**Comment by no-reply** [comment]
@jlhardes @samvera/hyrax-repo-managers: there's some discussion about the scope of permissions that are needed here.

it seems to be true that the current embargo and lease controllers don't bother checking for authentication at all--which seems like a serious issue.

@maxkadel's proposed solution in #4789 is to limit embargo editing to admin users (it sounds like the client on DCE's side is advocating that approach). for other permissions, i think we allow users with edit access to the relevant object to edit the permission/ACL---that could be an alternative here?

who should be able to edit existing embargoes/leases in a default Hyrax?

**Comment by maxkadel** [comment]
I agree that the fact that the embargo and lease controllers don't check permissions is the central issue here, and that probably my approach of making it admin-only is over-generalizing. 

I think I'm having trouble articulating the issue as a "rails engine" issue, and I've shared the particular use-case in which this has come up that highlights the issue. 

To me, since there are parts of the code that scope all the way down to `Hydra::AccessControls::Embargo` and `Hydra::AccessControls::Lease` it would make sense to have the controllers check against these qualities.

I'm not sure with CanCanCan what this would look like for the following examples:

* A user who can edit a specific work, and who can edit that object's Embargo
* A user who can edit a specific work, and who cannot edit any Embargoes
* A user who can edit any and all works, but not their Embargoes
* A user who can edit all works and their Embargoes

**Comment by no-reply** [comment]
@maxkadel thanks, this is a helpful comment.

there's a part of this that seems straightforward: the embargo and lease controllers provided by hyrax should enforce application `::Ability` on controller actions. i think that's true whatever hyrax's default abilities are or should be.

thinking about it this way, i can't see any reason not to move forward with that part of this change right away. it would be better to be more conservative/restrictive rather than less, and that would be the impact of introducing the authorization checks. since (as far as i can tell) only admins are ever actually linked to these controller actions, being overly restrictive while we sort out the rest isn't likely to impact anyone negatively, but would give applications a control over their own permissions that has been missing here to date.

**Comment by jlhardes** [comment]
For default Hyrax behavior it makes sense that if you can edit a work you can edit the embargo or lease on that work. So someone who has access to edit specific works can make changes to embargoes and leases on those specific works and someone who has access to edit all works (admin user) can make changes to embargoes and leases on all works.

I am thinking that setting things up for default Hyrax to account for the other two cases from @maxkadel:

- A user who can edit a specific work, and who cannot edit any Embargoes
- A user who can edit any and all works, but not their Embargoes

should be something that can be configured with documentation on how to do that configuration but not necessarily be actions that are ready to go in Hyrax by default. 

Does that help scope this or does this complicate things more based on how things are working now?

**Comment by no-reply** [comment]
an update: i took a closer look at this and discovered that there's an existing test for authorization that rejects and redirects unless the user has edit access on the OBJECT that the embargo/lease applies to. 
https://github.com/samvera/hyrax/blob/801eaa2dbdb6e581a07cf08e39eb601373101465/spec/controllers/hyrax/embargoes_controller_spec.rb#L30-L36

the implementation for this is in `ManagesEmbargoes` https://github.com/samvera/hyrax/blob/801eaa2dbdb6e581a07cf08e39eb601373101465/app/controllers/concerns/hyrax/manages_embargoes.rb#L9.

with this new information, the behavior i think is in place now seems like a reasonable default. it would be fantastic if someone could replicate this via UI:

  - given a Work, when i try to browse directly to its embargo/lease edit page:
    - as an unauthenticated user, i get an authorization error;
    - as a new user, without edit access for the work, i get an authorization error;
    - as the depositor, or other user with edit access, i can edit the embargo.

even if this is all correct and observed behavior, i think the implementation is wrong: the controller should explicitly check the authorization for the SPECIFIC embargo/lease object, and the application's `::Ability` should handle the decision about why and how that authorization is allowed.

**Comment by maxkadel** [comment]
From current master:

1. Sign in as basic_user@example.com
2. Upload a "New Generic Work" with visibility "public"
3. Go directly to Embargoes edit path, still as basic_user  (http://localhost:3000/embargoes/3484zg88m/edit) - can open
4. Attempt to log out, takes me to home page and get flash message that "User does not exist" and don't get signed out, which seems like a different question
5. Go into an incognito window, where I am an anonymous user, and go to http://localhost:3000/embargoes/3484zg88m/edit and get "You are not authorized to access this page"
6. Log in as admin@example.com and go to http://localhost:3000/embargoes/3484zg88m/edit, get "You are not authorized to access this page" (I don't know if the admin role is really created on a clean install, though?)
7. Log in as another_user@example.com and go to http://localhost:3000/embargoes/3484zg88m/edit, get "You are not authorized to access this page"

**Comment by no-reply** [comment]
@maxkadel thanks.

i'm deescalating this from `bug` with this new info in hand. i think there's some small technical work to do now, and some discussion about desired functionality, but the default behavior looks like it's as-intended.
